core/src/banking_stage/transaction_scheduler/in_flight_tracker.rs

```diff
@@ -0,0 +1,142 @@
+use {
+    super::thread_aware_account_locks::ThreadId,
+    crate::banking_stage::scheduler_messages::TransactionBatchId, std::collections::HashMap,
+};
+
+/// Tracks the number of transactions that are in flight for each thread.
+pub struct InFlightTracker {
+    num_in_flight_per_thread: Vec<usize>,
+    cus_in_flight_per_thread: Vec<u64>,
+    batches: HashMap<TransactionBatchId, BatchEntry>,
+}
+
+struct BatchEntry {
+    thread_id: ThreadId,
+    num_transactions: usize,
+    total_cus: u64,
+}
+
+impl InFlightTracker {
+    pub fn new(num_threads: usize) -> Self {
+        Self {
+            num_in_flight_per_thread: vec![0; num_threads],
+            cus_in_flight_per_thread: vec![0; num_threads],
+            batches: HashMap::new(),
+        }
+    }
+
+    /// Returns the number of transactions that are in flight for each thread.
+    pub fn num_in_flight_per_thread(&self) -> &[usize] {
+        &self.num_in_flight_per_thread
+    }
+
+    /// Returns the number of cus that are in flight for each thread.
+    pub fn cus_in_flight_per_thread(&self) -> &[u64] {
+        &self.cus_in_flight_per_thread
+    }
+
+    /// Add a new batch with given `batch_id` and `num_transactions` to the
+    /// thread with given `thread_id`.
+    ///
+    /// # Panics
+    /// Panics if the batch id is already being tracked.
+    pub fn track_batch(
+        &mut self,
+        batch_id: TransactionBatchId,
+        num_transactions: usize,
+        total_cus: u64,
```

Comment:
Assume the CUs here are from compute_budget.cu_limit (aka requested_cu_limit). Due to https://github.com/solana-labs/solana/issues/31379 still open, probably worth pointing out the CUs will not be close "actual consumed CUs", using it to load-balance would unreliable to the least, or potentially be gamed to the worst.

Reply:
Yeah, initially i'm just using the number of txs for load balancing because requested CUs aren't trustworthy currently...

Reply:
right! But if, that's a big if, transactions are charged with requested-cu, then scheduler can still use requested-cu regardless how inaccurate it is 🤑